src/Microsoft.AspNet.OData/Extensions/HttpRequestMessageExtensions.cs

```diff
@@ -410,6 +412,57 @@ public static IEnumerable<IODataRoutingConvention> GetRoutingConventions(this Ht
 
             return request.GetRequestContainer().GetServices<IODataRoutingConvention>();
         }
+        /// <summary>
+        /// Checks whether the request is a POST targeted at a resource path ending in /$query.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="oDataPath">The OData path.</param>
+        /// <returns>true if the request path has $query segment.</returns>
+        internal static bool IsQueryRequest(this HttpRequestMessage request, string oDataPath)
+        {
+            return request.Method.Equals(HttpMethod.Post) && 
+                oDataPath?.TrimEnd('/').EndsWith('/' + ODataRouteConstants.QuerySegment, StringComparison.OrdinalIgnoreCase) == true;
+        }
+
+        /// <summary>
+        /// Transforms a POST request targeted at a resource path ending in $query into a GET request. 
+        /// The query options are parsed from the request body and appended to the request URL.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        /// <param name="requestContainerFactory">Request container factory.</param>
+        internal static void TransformQueryRequest(this HttpRequestMessage request, Func<IServiceProvider> requestContainerFactory)
+        {
+            if (requestContainerFactory == null)
+            {
+                throw Error.ArgumentNull("requestContainerFactory");
+            }
+
+            IServiceProvider requestContainer = requestContainerFactory();
+
+            // Fetch parsers available in the request container for parsing the query options in the request body
+            IEnumerable<IODataQueryOptionsParser> queryOptionsParsers = requestContainer.GetRequiredService<IEnumerable<IODataQueryOptionsParser>>();
```

Comment:
>requestContainer.GetRequiredService<IEnumerable<IODataQueryOptionsParser>> [](start = 72, length = 74)

I would rather see ODataQueryOptionsParserFactory as an instantiable class that has a GetQueryOptionParser(request) method that returns the query option parser. I could pass my custom query option parser, or the default query option parser, to the constructor or add my custom query option parser to that class. Lines 444-454 could be encapsulated within that method.

Reply:
@mikepizzo Here's how I anticipated plugging in of a custom query options parser to work
https://github.com/OData/WebApi/pull/2191/files#diff-9e480fb8b0555db1128de67981ff5e0aR305
I'd like to know if your idea about making the class instantiatable was primarily to support plugging in of parsers or whether my approach suffices.
I added the `GetQueryOptionsParser(request)` method to the factory method

Reply:
Slightly different that I had envisioned it, but I think this is fine.  I liked having a strongly typed class/interface passed in to the service container, but IEnumerable<IODataQueryOptionsParser> works fine.

---
In reply to: [488835406](https://github.com/OData/WebApi/pull/2191#discussion_r488835406) [](ancestors = 488835406)